framework/src/systems/NonlinearSystemBase.C

```diff
@@ -3103,3 +3134,73 @@ NonlinearSystemBase::mortarJacobianConstraints(bool displaced)
     it->second();
   }
 }
+
+void
+NonlinearSystemBase::computeInitialJacobian(NonlinearImplicitSystem & sys)
+{
+#ifdef LIBMESH_HAVE_PETSC
+  if (dynamic_cast<PetscMatrix<Real> *>(sys.matrix))
+  {
+    auto & petsc_matrix = *dynamic_cast<PetscMatrix<Real> *>(sys.matrix);
+    _computing_initial_jacobian = true;
+    _fe_problem.computeJacobianSys(sys, *_current_solution, *sys.matrix);
+    _computing_initial_jacobian = false;
+
+    // container for repeated access of element global dof indices
+    std::vector<dof_id_type> dof_indices;
+
+    auto & field_variables = _vars[0].fieldVariables();
+    auto & scalar_variables = _vars[0].scalars();
+
+    std::vector<Real> inverse_scaling_factors(field_variables.size() + scalar_variables.size(), 0);
+    auto & dof_map = dofMap();
+
+    // Compute our scaling factors for the spatial field variables
+    for (const auto & elem : *mesh().getActiveLocalElementRange())
+    {
+      for (MooseIndex(field_variables) i = 0; i < field_variables.size(); ++i)
+      {
+        auto & field_variable = *field_variables[i];
+        dof_map.dof_indices(elem, dof_indices, field_variable.number());
+        for (auto dof_index : dof_indices)
+          if (dof_map.local_index(dof_index))
+          {
+            // For now we will use the diagonal for determining scaling
+            auto mat_value = petsc_matrix(dof_index, dof_index);
```

Comment:
It might be expensive since we need to use MatGetRow and also need to search for column. 

It is better to grab the diagonal matrix.

Minor thing: we need to check if an index is locally owned

Reply:
Done